videostore/utils.go

```diff
@@ -229,41 +263,85 @@ func formatDateTimeToString(dateTime time.Time) string {
 
 // matchStorageToRange returns a list of files that fall within the provided time range.
 // Includes trimming video files to the time range if they overlap.
-func matchStorageToRange(files []string, start, end time.Time, logger logging.Logger) []string {
+// The matcher assumes that the input files list is sorted by start time and the underylying
+// video segments do not overlap.
+func matchStorageToRange(files []fileWithDate, start, end time.Time, logger logging.Logger) []string {
 	var matchedFiles []string
-	for _, file := range files {
-		dateTime, err := extractDateTimeFromFilename(file)
-		if err != nil {
-			logger.Debugf("failed to extract datetime from filename: %s, error: %v", file, err)
-			continue
+	// Cache of the first matched video file's width, height, and codec
+	// to ensure every video in the matched files set have the same params.
+	var firstSeenVideoInfo videoInfo
+	// Find the first file to consider for matching. First search for the first file that starts after the query start time.
+	// We then want to consider the previous file as well, since it may overlap with the query start time.
+	var firstFileIndex int
+	firstFileIndexFound := false
+	for i, file := range files {
+		if file.date.After(start) {
+			firstFileIndex = i - 1
+			firstFileIndexFound = true
+			break
+		}
+	}
```

Comment:
Not sure if it's necessary, but there is a `slices.BinarySearch` we could use if it would improve performance, given we already have a sorted slice.

Reply:
what if we binary search to find the index that is before the start time and the index that is after the end time & only iterate over those elements?